### ISSUE TITLE
Host downloaded data files (spacy.en) remotely?

### ISSUE THREAD

**Issue by bittlingmayer** [explosion/spaCy#443]
I'm experimenting with running spaCy on AWS Lambda, which has a total package size limit of 250MB.

Is there an easy way to decouple the lib from the (ca 500 MB) data, ie to point spaCy at a data file hosted on remote storage, for example on S3?

**Comment by syllog1sm** [comment]
Does ```nlp = English(data_dir='some/path/')``` work?

**Comment by bittlingmayer** [comment]
Yes and no. :-)

I can run `python -m spacy.en.download` in the init, and the container ie contents of /tmp (max 512MB) persist if it's used against soon thereafter, which means init latency is no issue, but it seems like asking for trouble to start so close to the limit, and also it will be impossible to do anything involving multiple languages.

https://github.com/spacy-io/sputnik/blob/master/sputnik/dir_package.py seems only to handle the filesystem, but I can't point it at remote URLs.  (I could modify it for our use, but I would want your opinion before I did that.)

I will play a bit more.

**Comment by bittlingmayer** [comment]
Reporting back

I did try writing the data to /tmp:
> sputnik.install('spacy', spacy.about.__version__, 'en_default', data_path='/tmp/spacy/')

but hit that limit as expected:

> module initialization error: [Errno 28] No space left on device

So the only options I see are:

1. modify Sputnik to support remote storage like S3
2. trim the model
3. load the model into memory (joking)
4. forget this idea

Re 2 (trimming), do I need all of the following?

>  ['en_default-1.0.3', 'en_default-1.0.8', 'de-1.0.0', 'en-1.0.9', 'en_default-1.0.0', 'en_default-1.0.5', 'en_default-1.0.4', 'en_glove_cc_300_1m_vectors-1.0.0', 'en_default-1.0.6', 'en-1.1.0']

And is there an elegant way to download only one?

**Comment by asafge** [comment]
@bittlingmayer Hey, did you make any progress in this effort? Thanks

**Comment by bittlingmayer** [comment]
@asafge No sir, I went with option 4 for now.  I need it for a few languages and more down the road (so the total size would again increase), so just hitting a another machine or service for parsing is the best way to do this from a Lambda.

**Comment by asafge** [comment]
@bittlingmayer Yeah ended up doing something similar. I'm using heroku but they also have storage limitations. Thanks for responding!

**Comment by anton164** [comment]
Hey @asafge @bittlingmayer, could you recommend a service where I can host a python script running the spacy model (need en_core_web_md, so roughly 1GB, but ideally up to 2GB)? I've ran into the max slug size limit on Heroku..

**Comment by tkaria** [comment]
I'm encountering the same issue with AWS lambda and the size limit. One of the potential issues I noticed while zipping the file (`zip -r9 $VIRTUAL_ENV/lib/python2.7/site-packages`) is that the symlink created in the `spacy/data` directory seems to get re-added by the symlink. But that's more zip behavior than spacy. Anyone have any luck running spacy from lambda?

**Comment by honnibal** [comment]
You can either run `en_core_web_sm`, which is 50mb, or you can run `spacy-nightly` -- the model for v2 is smaller and loads much faster, and has better accuracy. The disadvantage is that it's slower, especially with small batch sizes. If you can batch your inputs, it's about 2x slower on CPU. If you have to process texts one-by-one, it's about 10x slower.

**Comment by rafapetter** [comment]
I'm facing the same problem. 

Using zappa to upload to AWS lambda a very simple flask one-file project, with spacy and the lightest en model from nightly. Overall project zip size is about 70MB. Everything works fine on local, but on AWS lambda I keep getting "'NoneType' object is not callable".

I doubt this is a spacy problem. So my question here is what are the alternatives? Is there a way to exclude spacy+model from the zappa zip file, and import them dynamically when the lambda executes?

**Comment by tkaria** [comment]
That doesn't sound like a spacy problem in particular. It's a pretty generic error. The issue I was having was lambda specific about size of the unzipped contents being larger than 250mb. I would try to debug locally. Also, are you using event triggers (async) lambdas? Because it could be the structure of the event you're receiving. Happy to help if I can but let's not pollute the issue tracker with this. Email me directly.

**Comment by rafapetter** [comment]
@tkaria Ok, let me rephrase the question: Is there a lighter version of the spacy lib if I just want to do sentence similarity? If not, any suggestion on how to load spacy only at runtime?
